### ISSUE TITLE
关于使用dp=5时，精度没达到原本预期的问题

### ISSUE THREAD

**Issue by sherlock-CHN** [wenzhu23333/Differential-Privacy-Based-Federated-Learning#23]
作者你好
我在使用高斯机制，dp=5，mnist数据集的时候，并没有达到你原本在log里保存的dp=5那么高的精度
目前在我的机器中运行dp=5时，精度10-20%左右
在您保存的训练数据中，dp=5最后的精度可达到54%
请问是哪里错误了吗？
图1是我正在训练时的精度，图2是您保存的训练数据精度
![image](https://github.com/wenzhu23333/Differential-Privacy-Based-Federated-Learning/assets/54991793/448b7510-1d7d-4aaf-a625-e1326a0f3de4)

![image](https://github.com/wenzhu23333/Differential-Privacy-Based-Federated-Learning/assets/54991793/b2ddf8eb-6125-4cc5-b252-9df232374960)

**Comment by wenzhu23333** [comment]
请提供你运行该脚本的执行命令，核对参数是否和run.sh中相同

**Comment by sherlock-CHN** [comment]
`def args_parser():
    parser = argparse.ArgumentParser()
    # federated arguments
    parser.add_argument('--epochs', type=int, default=100, help="rounds of training")
    parser.add_argument('--num_users', type=int, default=100, help="number of users: K")
    parser.add_argument('--frac', type=float, default=0.5, help="the fraction of clients: C")
    parser.add_argument('--bs', type=int, default=1024, help="test batch size")
    parser.add_argument('--lr', type=float, default=0.1, help="learning rate")
    parser.add_argument('--lr_decay', type=float, default=0.995, help="learning rate decay each round")

    # model arguments
    parser.add_argument('--model', type=str, default='cnn', help='model name')

    # other arguments
    parser.add_argument('--dataset', type=str, default='mnist', help="name of dataset")
    parser.add_argument('--iid', action='store_true', help='whether i.i.d or not')
    parser.add_argument('--num_classes', type=int, default=10, help="number of classes")
    parser.add_argument('--num_channels', type=int, default=1, help="number of channels of imges")
    parser.add_argument('--gpu', type=int, default=0, help="GPU ID, -1 for CPU")

    parser.add_argument('--dp_mechanism', type=str, default='Gaussian',
                        help='differential privacy mechanism')
    parser.add_argument('--dp_epsilon', type=float, default=5,
                        help='differential privacy epsilon')
    parser.add_argument('--dp_delta', type=float, default=1e-5,
                        help='differential privacy delta')
    parser.add_argument('--dp_clip', type=float, default=10,
                        help='differential privacy clip')
    parser.add_argument('--dp_sample', type=float, default=1, help='sample rate for moment account')

    parser.add_argument('--serial', action='store_true', help='partial serial running to save the gpu memory')
    parser.add_argument('--serial_bs', type=int, default=128, help='partial serial running batch size')

    args = parser.parse_args()
    return args`
   
只修改了dp_epsilon=5，然后直接运行main.py文件

**Comment by wenzhu23333** [comment]
已修复，默认frac应为0.2，之前改代码的时候忘记了

**Comment by sherlock-CHN** [comment]
感谢感谢，但我现在使用Femnist和cifar训练的效果很差，可以分享您之前设置的参数吗

**Comment by wenzhu23333** [comment]
以下是我曾经跑的时候使用的命令，可以参考下：
`python3 -u main.py --dataset femnist --lr 0.2 --lr_decay 0.995 --frac 1 --gpu 2 --dp_epsilon 30`

`def args_parser():
    parser = argparse.ArgumentParser()
    # federated arguments
    parser.add_argument('--epochs', type=int, default=500, help="rounds of training")
    parser.add_argument('--num_users', type=int, default=100, help="number of users: K")
    parser.add_argument('--frac', type=float, default=0.2, help="ction of clients: C")
    parser.add_argument('--local_ep', type=int, default=1, help="the number of local epochs: E")
    parser.add_argument('--local_bs', type=int, default=1024, help="local batch size: B")
    parser.add_argument('--bs', type=int, default=1024, help="test batch size")
    parser.add_argument('--lr', type=float, default=0.1, help="learning rate")
    parser.add_argument('--lr_decay', type=float, default=0.995, help="learning rate decay each round")
    parser.add_argument('--momentum', type=float, default=0.5, help="SGD momentum (default: 0.5)")
    parser.add_argument('--split', type=str, default='user', help="train-test split type, user or sample")

    # model arguments
    parser.add_argument('--model', type=str, default='cnn', help='model name')

    # other arguments
    parser.add_argument('--dataset', type=str, default='mnist', help="name of dataset")
    parser.add_argument('--iid', action='store_true', help='whether i.i.d or not')
    parser.add_argument('--num_classes', type=int, default=10, help="number of classes")
    parser.add_argument('--num_channels', type=int, default=1, help="number of channels of imges")
    parser.add_argument('--gpu', type=int, default=0, help="GPU ID, -1 for CPU")

    parser.add_argument('--dp_mechanism', type=str, default='Gaussian',
                        help='differential privacy mechanism')
    parser.add_argument('--dp_epsilon', type=float, default=50.0,
                        help='differential privacy epsilon')
    parser.add_argument('--dp_delta', type=float, default=1e-5,
                        help='differential privacy delta')
    parser.add_argument('--dp_clip', type=float, default=15,
                        help='differential privacy clip')

    parser.add_argument('--tau', type=float, default=1, help='niss tau')
    parser.add_argument('--unit_noise_scale', type=float, default=0.0001, help='niss unit noise')

    parser.add_argument('--collusion', type=float, default=0.5)

    parser.add_argument('--scheme', type=str, default='no_dp')

    parser.add_argument('--serial', help='using opacus', action='store_true')
    parser.add_argument('--serial_bs', type=int, default=64, help='serial batch size')

    args = parser.parse_args()
    return args`
